File: using_Go/8_myAtoi.go
var daf = make (map[string] []string)
var state string
var ans int 
var sign int
var INT_MAX int
var INT_MIN int
func get_col(c string)int{
    if (c==" ") {return 0}
	if (c=="+" || c =="-") {return 1}
	if (c>="0" && c<="9") {return 2}
	return 3
}

func get(c string){
    state = daf[state][get_col(c)]
    if (state=="in_num"){
        temp,_:= strconv.Atoi(c)
        ans=ans*10 + temp
        if (ans>INT_MAX){ans=INT_MAX}
        if (ans<INT_MIN){ans =INT_MIN}
    }
    if (state=="signed"){
        if (c=="+"){
            sign=1
        }
        if(c=="-"){
            sign=-1
        }
    }
}
func myAtoi(s string) int {
    daf["start"] = []string{"start","signed","in_num","end"}
    daf["signed"] = []string{"end","end","in_num","end"}
    daf["in_num"] = []string{"end","end","in_num","end"}
    daf["end"] = []string{"end","end","end","end"}
    state="start"
    ans =0
    sign =1
    INT_MIN = -2147483648
    INT_MAX = 2147483647
    for _,c :=range s{
        get(string(c))
    }
    if ((sign*ans)<INT_MIN){return INT_MIN}
    if ((sign*ans)>INT_MAX) {return INT_MAX}
    return sign*ans
}

File: using_python/8_myAtoi.py
INT_MAX = 2**31 - 1
INT_MIN = -2**31


class Automaton:
    def __init__(self):
        self.state = "start"
        self.sign = 1
        self.ans = 0
        self.table = {
            "start": ["start", "signed", "in_number", "end"],
            'signed': ['end', 'end', 'in_number', 'end'],
            'in_number': ['end', 'end', 'in_number', 'end'],
            'end': ['end', 'end', 'end', 'end'],
        }

    def get_col(self, c):
        if c.isspace():
            return 0
        if c == "+" or c == "-":
            return 1
        if c.isdigit():
            return 2
        return 3

    def get(self, c):
        self.state = self.table[self.state][self.get_col(c)]
        if self.state == "in_number":
            self.ans = self.ans * 10 + int(c)
            self.ans = min(self.ans, INT_MAX) if self.sign == 1 else min(
                self.ans, -INT_MIN)
        elif self.state == 'signed':
            self.sign = 1 if c == '+' else -1


class Solution():
    def myAtoi(self, str: str) -> int:
        automaton = Automaton()
        for c in str:
            automaton.get(c)
        return automaton.sign * automaton.ans

    def myAtoi2(self, str: str) -> int:
        flag_start = False
        sign = 1
        ans = 0
        for c in s:
            if c == " " and not flag_start:
                continue
            elif c == "+" and not flag_start:
                flag_start = True
            elif c == "-" and not flag_start:
                flag_start = True
                sign = -sign
            elif c.isdigit() and not flag_start:
                flag_start = True
                ans = ans * 10 + int(c)
            elif c.isdigit() and flag_start:
                ans = ans * 10 + int(c)

            else:
                if INT_MIN <=sign * ans <=INT_MAX:
                    return sign * ans
                elif sign * ans<INT_MIN:
                    return INT_MIN
                else: return INT_MAX
        
        if INT_MIN <=sign * ans <=INT_MAX:
                    return sign * ans
        elif sign * ans<INT_MIN:
                    return INT_MIN
        else: return INT_MAX


so = Solution()
print(so.myAtoi2("   +456464 loss"))
